spdlog/src/formatter/pattern_formatter/mod.rs

```diff
@@ -0,0 +1,1445 @@
+//! This module provides a pattern-string based formatter.
+//!
+//! The [`PatternFormatter`] struct defines a formatter that formats log
+//! messages against a specific text pattern.
+//!
+//! Patterns are represented by the [`Pattern`] trait. You can create your own
+//! pattern by implementing the [`Pattern`] trait.
+//!
+//! You can also build a pattern with the [`pattern`][pattern-macro] macro.
+//!
+//! [pattern-macro]: crate::pattern
+
+pub mod patterns;
+
+use std::{fmt::Write, ops::Range, sync::Arc};
+
+use crate::{
+    formatter::{FmtExtraInfo, FmtExtraInfoBuilder, Formatter},
+    Error, Record, StringBuf,
+};
+
+#[allow(missing_docs)]
+pub mod macros {
+    pub use ::spdlog_macros::pattern as pattern_impl;
+}
```

Comment:
The point is that procedural macros are rendered with less information in the doc than macros by example.

So if we simply re-export `spdlog_macros::pattern` from `spdlog`, it will be rendered as follows:

```
pattern!() { /* proc-macro */ }
```

As a comparison, if we define a new macro `spdlog::pattern` and forward its arguments to `spdlog_macros::pattern`, `pattern` is rendered as follows:

```
macro_rules! pattern {
    ( $($t:tt)* ) => { ... };
}
```

To be honest, `rustdoc` renders procedural macros in a quite misleading way. I could be thinking that `pattern!` takes no arguments, which is not the case. A macro by example is more clearer.

So I made a `pattern!` wrapper in `spdlog` which simply forwards its arguments to `spdlog_macros::pattern`. The problem is we cannot simply import `spdlog_macros::pattern` because of name collision. Thus `spdlog_macros::pattern` is imported in `spdlog` with name `pattern_impl`. Placing such an item, which is completely implementation details, pollutes the doc. So I place it in a new undocumented module named `macros`.

Reply:
It's actually documented. You probably want `#[doc(hidden)]` instead of `#[allow(missing_docs)]`. 😇

```rust
#[doc(hidden)]
pub use ::spdlog_macros::pattern as __pattern_impl;
```

Reply:
Wow, this problem is quite tricky.

My intention is to export the macro only to the `formatter` module and not to the root module, and consider your thoughts, a more detailed information is definitely better.

However, it looks like we don't have a clean way to do both, see [this comment](https://users.rust-lang.org/t/how-to-namespace-a-macro-rules-macro-within-a-module-or-macro-export-it-without-polluting-the-top-level-namespace/63779/5#answering-the-original-threads-title-xy-problem-_quid_-of-macro_exported-macros-1).

[Declarative macros 2.0](https://github.com/rust-lang/rust/issues/39412) will eventually solve such visibility issues. For now, I'm considering whether we should use the js / css hack mentioned in the comment. This may be the only way to do both purposes. 🥲

Reply:
A cleaner way, we just re-export it directly and then use doctest to tell users the form it accepts.

Pseudo code:

```rust
/// Build a pattern from a compile-time pattern template string.
///
/// It accepts input of the form:
///
/// ```
/// macro_rules! pattern {
///     ( $pattern:literal $(, {$custom_pattern:ident} => $custom_new:expr)* $(,)? ) => {
///         // ...
///         # spdlog::formatter::pattern!($pattern $(, {$custom_pattern} => $custom_new)*)
///     }
/// }
/// # pattern!("text only");
/// # pattern!("with pattern {loc}");
/// # pattern!(
/// #     "with custom pattern {custom_pattern}",
/// #     {custom_pattern} => MyPattern::new
/// # );
/// # pattern!(
/// #     "with custom multiple patterns {custom_pattern1} {custom_pattern2}",
/// #     {custom_pattern1} => MyPattern::new,
/// #     {custom_pattern2} => MyPattern::new
/// # );
/// ```
///
/// # Basic Usage
///
/// ... *Snipped*
pub use ::spdlog_macros::pattern;
```

BTW, I changed `$($t:tt)*` to `$pattern:literal $(, {$custom_pattern:ident} => $custom_new:expr)* $(,)?`, which provides users with more detailed information.

It looks like this:

![image](https://user-images.githubusercontent.com/33660554/177011172-daa642aa-1df6-479b-9c86-50be757fe940.png)

Now we have 3 options :)

Reply:
Maybe it can be made more clearer by this:

```rust
macro_rules! pattern {
    ( $t:literal $(,)? ) => {
        // ...
    };
    ( $t:literal, $( {$custom_pattern:ident} => $custom_new:expr ),+ $(,)? ) => {
        // ...
    }
}
```

`rustdoc` renders it like this:

![图片](https://user-images.githubusercontent.com/29401356/177022052-c119366f-815d-4a8a-9f92-c431e722d882.png)